File: ZAlgoDemo/ZAlgoDemo/ZAlgoLinkedList/ZAlgoList.m
//
//  ZAlgoList.m
//  ZAlgoDemo
//
//  Created by zjixin on 2019/7/19.
//  Copyright © 2019 zjixin. All rights reserved.
//

#import "ZAlgoList.h"
#import "stdlib.h"

#pragma mark -- ZAlgoListNode

BOOL hasCycle(struct ZAlgoListNode *head) {
    struct ZAlgoListNode *slow = head;
    struct ZAlgoListNode *fast = head;
    
    while (slow != NULL && fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
        if (fast == slow) {
            return true;
        }
    }
    return false;
}

#pragma mark -- ZAlgoList

void appendToTail(struct ZAlgoList *list, int val) {
    if (list == NULL) {
        return;
    }
    
    if ((list->head) == NULL) {
        struct ZAlgoListNode *node = malloc(sizeof(struct ZAlgoListNode));
        node->val = val;
        node->next = nil;
        list->tail = node;
        list->head = node;
    } else {
        struct ZAlgoListNode *node = malloc(sizeof(struct ZAlgoListNode));
        node->val = val;
        node->next = nil;
        list->tail->next = node;
        list->tail = node;
    }
}

void appendNodeToTail(struct ZAlgoList *list, struct ZAlgoListNode *node) {
    if (list == NULL) {
        return;
    }
    
    if (list->head == NULL) {
        list->tail = node;
        list->head = node;
    } else {
        list->tail->next = node;
        list->tail = node;
    }
    list->tail->next = nil;
}

void displayAllNodeValue(struct ZAlgoListNode *head) {
    struct ZAlgoListNode *node = head;
    while (node != NULL) {
        printf("%d->", node->val);
        node = node->next;
    }
    printf("\n");
}

int _calculateLinkedListLength(struct ZAlgoListNode *head) {
    int count = 0;
    struct ZAlgoListNode *node = head;
    while (node != NULL) {
        count++;
        node = node->next;
    }
    return count;
}

struct ZAlgoListNode *getIntersectionNode(struct ZAlgoListNode *headA, struct ZAlgoListNode *headB) {
    int lengthA = _calculateLinkedListLength(headA);
    int lengthB = _calculateLinkedListLength(headB);
    
    int result = lengthA - lengthB;
    
    struct ZAlgoListNode * nodeA = headA;
    struct ZAlgoListNode * nodeB = headB;
    
    for (int i = 0; i < abs(result); i++) {
        if (result > 0) {
            nodeA = nodeA->next;
        } else {
            nodeB = nodeB->next;
        }
    }
    
    while (nodeA && nodeB) {
        if (nodeB == nodeA) {
            return nodeA;
        }
        
        nodeA = nodeA->next;
        nodeB = nodeB->next;
    }
    return NULL;
}

struct ZAlgoListNode *reverseList(struct ZAlgoListNode *head) {
    if (head == NULL) {
        return NULL;
    }
    
    struct ZAlgoListNode *preNode = NULL;
    struct ZAlgoListNode *curNode = head;
    
    while (curNode != NULL) {
        struct ZAlgoListNode * nextNode = curNode->next;
        
        curNode->next = preNode;
        preNode = curNode;
        curNode = nextNode;
    }
    return preNode;
}

struct ZAlgoListNode *mergeTwoLists(struct ZAlgoListNode *headA, struct ZAlgoListNode *headB){
    struct ZAlgoList *list = malloc(sizeof(struct ZAlgoList));
    
    struct ZAlgoListNode *curHeadA = headA;
    struct ZAlgoListNode *curHeadB = headB;
    
    while (curHeadA != NULL || curHeadB != NULL) {
        if (curHeadA != NULL && curHeadB == NULL) {
            appendNodeToTail(list, curHeadA);
            break;
            
        } else if (curHeadA == NULL && curHeadB != NULL) {
            appendNodeToTail(list, curHeadB);
            break;
            
        } else {
            if (curHeadA->val <= curHeadB->val) {
                appendToTail(list, curHeadA->val);
                curHeadA = curHeadA->next;
            } else {
                appendToTail(list, curHeadB->val);
                curHeadB = curHeadB->next;
            }
            
        }
    }
    return list->head;
}


struct ZAlgoListNode *mergeTwoListsWithRecursion(struct ZAlgoListNode *headA, struct ZAlgoListNode *headB) {
    if (headA == NULL) {
        return headB;
    }
    
    if (headB == NULL) {
        return headA;
    }
    
    struct ZAlgoListNode *resultHead = malloc(sizeof(struct ZAlgoListNode));
    
    if (headA->val <= headB->val) {
        resultHead->val = headA->val;
        resultHead->next = mergeTwoListsWithRecursion(headA->next, headB);
    } else {
        resultHead->val = headB->val;
        resultHead->next = mergeTwoListsWithRecursion(headA, headB->next);
    }
    return resultHead;
}

struct ZAlgoListNode *findKthNodeToTail(struct ZAlgoListNode *head, int k) {
    if (head == NULL) {
        return NULL;
    }
    
    struct ZAlgoListNode *pAhead = head;
    struct ZAlgoListNode *pBehind = NULL;
    
    for (int i = 0; i < k - 1; i++) {
        if (pAhead->next != NULL) {
            pAhead = pAhead->next;
        } else {
            return NULL;
        }
    }
    pBehind = head;
    while (pAhead->next != NULL) {
        pAhead = pAhead->next;
        pBehind = pBehind->next;
    }
    return pBehind;
}

struct ZAlgoListNode *removeNthFromEnd(struct ZAlgoListNode *head, int n) {
    if (head == NULL) {
        return NULL;
    }
    
    struct ZAlgoListNode *pAhead = head;
    struct ZAlgoListNode *pBehind = NULL;
    
    for (int i = 0; i < n; i++) {
        pAhead = pAhead->next;
    }
    
    if (pAhead == NULL) {
        return head->next;
    }
    
    pBehind = head;
    while (pAhead->next != NULL) {
        pAhead = pAhead->next;
        pBehind = pBehind->next;
    }
    pBehind->next = pBehind->next->next;
    return head;
}

void printListReversingly(struct ZAlgoListNode *head) {
    struct ZAlgoListNode *node = head;
    if (node != NULL) {
        printListReversingly(node->next);
        printf("%d->", node->val);
    }
}

File: ZAlgoDemoSwift/ZAlgoDemoSwift/ZAlgoDemoLinkedList/ZAlgoList.swift
//
//  ZAlgoList.swift
//  ZAlgoDemoSwift
//
//  Created by zjixin on 2019/6/25.
//  Copyright © 2019 zjixin. All rights reserved.
//

import Foundation

/// 单向链表节点
class ZAlgoListNode : NSObject {
    var val:Int
    var next:ZAlgoListNode?
    
    init(_ value:Int) {
        self.val = value
        self.next = nil
    }
    /*  快行指针
     两个指针访问链表，一个在前一个在后，或者一个移动快另一个移动慢，这就是快行指针
     */
    
    /// 判断一个链表是否有环
    ///
    /// - Returns: YES：包含环
    open func hasCycle() -> Bool {
        var slow: ZAlgoListNode? = self
        var fast: ZAlgoListNode? = self
        
        while ((fast != nil) && fast?.next != nil) {
            slow = slow!.next
            fast = fast!.next?.next
            if slow === fast {
                return true
            }
        }
        return false
    }
    
    open func _calculateLinkedListLength() -> Int {
        var count = 0
        var node: ZAlgoListNode? = self
        while node != nil {
            count += 1
            node = node?.next
        }
        return count
    }
    
    open func displayAllNodeValue() {
        var listNode: ZAlgoListNode? = self
        print(self)
        var log = "\t"
        while listNode != nil {
            log += "\(listNode!.val)->"
            listNode = listNode!.next
        }
        print(log)
    }
    
    open func printListReversingly() {
        let node: ZAlgoListNode? = self;
        if node != nil {
            if node!.next != nil {
                node?.next?.printListReversingly()
            }
            print("\(node!.val)")
        }
    }
}

/// 单向链表
class ZAlgoList: NSObject {
    public var head: ZAlgoListNode?
    public var tail: ZAlgoListNode?
    
    override init() {
        
    }
    
    /// 尾插法
    ///
    /// - Parameter val: 插入的value
    open func appendToTail(val: Int) {
        if tail == nil {
            tail = ZAlgoListNode.init(val)
            head = tail
        } else {
            tail!.next = ZAlgoListNode.init(val)
            tail = tail!.next
        }
    }
    
    open func appendNodeToTail(node: ZAlgoListNode?) {
        guard node != nil else {
            return
        }
        
        if tail == nil {
            tail = node
            head = tail
        } else {
            tail!.next = node
            tail = node;
        }
    }
    
    /// 头插法
    ///
    /// - Parameter val: 插入的value
    open func appendToHead(val:Int) {
        if head == nil {
            head = ZAlgoListNode.init(val)
            tail = head
        } else {
            let listNode = ZAlgoListNode.init(val)
            listNode.next = head
            head = listNode
        }
    }
    
    open func deleteTail() -> ZAlgoListNode? {
        if tail == nil {
            return nil
            
        } else {
            var newTailNode:ZAlgoListNode?
            let oldTailNode = self.tail
            var listNode = head
            
            while (listNode != nil) {
                if listNode!.next == self.tail {
                    newTailNode = listNode
                    break
                }
                listNode = listNode!.next
            }
            newTailNode?.next = nil
            self.tail = newTailNode
            return oldTailNode
        }
        
    }
    
    open func reverserList() -> ZAlgoListNode? {
        guard head != nil else {
            return nil
        }
        
        var preNode: ZAlgoListNode?
        var curNode: ZAlgoListNode? = head!
        while curNode != nil {
            let nextNode = curNode!.next
            
            curNode!.next = preNode
            preNode = curNode
            curNode = nextNode
        }
        return preNode
    }
    
    open func displayAllNodeValue() {
        var listNode = head
        print(self)
        var log = "\t"
        while listNode != nil {
            log += "\(listNode!.val)->"
            listNode = listNode!.next
        }
        print(log)
    }
    
    open func deleteDuplicates() -> ZAlgoListNode? {
        guard self.head != nil else {
            return nil;
        }
        
        if self.head!.next == nil {
            return self.head;
        }
        
        var curNode: ZAlgoListNode? = self.head
        while curNode?.next != nil {
            if curNode!.val == curNode!.next!.val {
                curNode!.next = curNode!.next?.next
            } else {
                curNode = curNode!.next;
            }
        }
        return self.head;
    }
}
